File: 78. Subsets.c
class Solution {
public:
    void dfs(vector<int>& temp, vector<vector<int>>& result, vector<int>& nums, int index) {  
        //if (index>=nums.size()) {  
		//	return;
        //} 
        for (int i=index; i<nums.size(); i++) {  
            temp.push_back(nums[i]);
            result.push_back(temp);
            dfs(temp,result, nums, i+1);  
            temp.pop_back();
        }  
    }  
    
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> temp;
        result.push_back(temp);
        dfs(temp, result, nums, 0);
        return result;
    }
};

File: Algorithms/Recursion & Backtracking/78. Subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if len(nums)==0:
            return nums

        result = []
        n = len(nums)

        def dfs(nums, subset, result):
            result.append(subset)
            for i in range(len(nums)):
                dfs(nums[i+1:], subset+[nums[i]], result)
        dfs(nums, [], result)
        return result

        
